files/migrations.py

```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/python3
+"""
+Runs migration scripts for this project
+
+Intended to be triggered by the %post rpm scriptlet rather than manually.
+"""
+
+import logging
+import subprocess
+import sys
+from pathlib import Path
+
+from systemd.journal import JournalHandler
+
+log = logging.getLogger("securedrop-updater-migrations")
+log.addHandler(JournalHandler(SYSLOG_IDENTIFIER="securedrop-updater-migrations"))
+log.setLevel(logging.INFO)
+
+
+class Version:
+    """
+    A sortable software version
+
+    Reads version strings, and sorts by major.minor.patch[.etc[.etc …]]
+    """
+
+    def __init__(self, version):
+        """
+        Read version numbers into lists of ints, discard anything that comes
+        after the first character that is neither a period or a digit
+        """
+        chunks = []
+        # Opting for minimal code over execution speed
+        for part in version.split("."):
+            digits = ""
+            for char in part.split():
+                if char.isdigit():
+                    digits += char
+                else:
+                    # Treat RCs, betas etc. as if it were the full release
+                    break
+            chunks.append(int(digits))
+        self.version = tuple(chunks)
+
+    def __str__(self):
+        return ".".join(str(v) for v in self.version)
+
+
+class Migration(Version):
+    """
+    A sortable runnable migration
+
+    Subclass of Version so that we can sort migrations just as easily as version strings
+    """
+
+    def __init__(self, path):
+        super().__init__(path.name.rsplit(".", 1)[0])
+        self.path = path
+
+    def run_and_update_version_file(self, version_file):
+        """
+        Run this migration
+        """
+        rc = None
+        try:
+            log.info(f"Running migration for {self}")
+            process = subprocess.Popen([str(self.path)])
```

Comment:
Very good question that I missed to think about when you originally commented… and you're right that's basically just organic growth, it would certainly make logging less icky as well. However I don't quite see a good way to do this yet, do you have any suggestions?

Reply:
Refactored to use `importlib` which seemed like the least worst option to continue to use the version file name convention.